### DISCUSSION TITLE
How will directives affect block deprecations for static blocks?

### DISCUSSION THREAD

**Discussion by fabiankaegy** [WordPress/gutenberg#53023] [Interactivity API]
Let me preface this by saying that for the exact reasons I'm outlining and asking about here, I am a proponent of dynamic blocks a lot of the time anyways. Looking at directives however, I'm not sure I can even grasp how they will actually integrate with the block deprecation system in static WordPress blocks. 

When a static block is saved to the DB it saves the entire markup (including any directives) to the DB. And any changes in this markup require the editor screen to be loaded again and the post to be re saved before the updates are applied. 

For any block that is built today, this mostly is a styling issue. Because it means your block can always exist in multiple different versions throughout the site. So you need to ensure that your styles still apply properly for all the different versions of the old deprecations.

With the directives however this now also affects all the changes to your interactivity. And the `view.js` file is not "versioned" in the same way. So if you make an update to change the functionality of your `view.js` file to change how an existing action for example works alongside a change to the markup. This will break the action for all the other old instances of the bock still living on the site.

When the interactivity was handled manually in a `view.js` file with either vanilla JS or any frontend framework you could add custom code to handle the various versions of the block on the site. And create different experiences. But I didn't quite see how this could work with the declarative approach when we have one `view.js` file that is intended to serve infinite deprecations at the same time.

Not to mention that if I understand it correctly you would need to add a new block deprecarion any time you change something about the interactivity layer. Which is definitely something that is usually undergoing a lot more change over time than the actual markup of a given component.

**Comment by luisherranz** [comment]
The Interactivity API doesn't address any of the shortcomings of the static blocks, and it is not within their goals to do so. It does, however, put the **focus on dynamic blocks**, to, in a way, try to compensate for those shortcomings.

So the same limitations that exist today for static blocks, will keep existing as they are. In the same way that the static block styles must support all versions, static block views must support all versions as well.

Static blocks are very important to keep saving the _content_ in the database: headings, paragraphs, images, etc. But just the _content_, not the entire site. In the era of PHP themes, the rest of the site was created using PHP files and it was perfectly fine. In the era of block themes, the rest of the site can still be created with PHP files (dynamic blocks). The fact that we are storing them as JSONified HTML comments is incidental to the necessity of storing them somehow. But dynamic blocks are not _inferior_ or _less correct_ than static blocks just because they don't save their markup in the database. This means that the need for static blocks outside of WordPress Core should be very small, if not absent, and therefore its developer experience is not so important.

To make static blocks more resilient, WordPress now ships with the HTML Tag Processor. People writing static blocks should start using it to inject attributes dynamically and diminish the need for markup changes and deprecations. Ideally, the content saved in the database should be the semantic representation of that block. There's no need to serialize anything that is WordPress-related. This includes other things, but also the injection of directives.

```html
<!-- wp:my-plugin/image -->
<figure>
  <img src="..." alt="...">
</figure>
<!-- /wp:my-plugin/image -->
```

```php
$p = new WP_HTML_Tag_Processor( $content );
$p->next_tag( array( 'tag_name' => 'IMG' ) );
$p->set_attribute( 'data-wp-on.mouseover', 'actions.myPlugin.zoomImage' );
```

The developer experience of adding a directive using the HTML Tag Processor is not as good as adding it directly in the markup. But again, as the Interactivity API treats dynamic blocks as first citizens where this is not needed, and the usage of interactive static blocks outside of WordPress Core should be minimal, I think this added complexity is ok. 

There is a sense in the WordPress community that creating static blocks is the default and _less advanced_ way, but that path leads to frustrations when people start discovering their shortcomings. We should revert that trend and start treating dynamic blocks as the default and _beginner friendly_. Then, teach them that static blocks exist for a specific use case.

Actually, we intend to teach and document the Interactivity API using mostly dynamic blocks. How to combine directives and static blocks will be a section on its own, hopefully with a disclaimer explaining the difference between static and dynamic blocks, and guidance on how to make the decision about which one to use. It should also include a recommendation to use the HTML Tag Processor to minimize the need for deprecations.

_By the way, most of this is, of course, my personal opinion, and anyone with different experience or expectations can disagree 🙂_

**Reply by fabiankaegy** [reply]
Thanks for sharing all that insight. That all makes a ton of sense to me and is similar to how I feel about it.

Like I stated at the beginning of the discussion. I'm a fan of dynamic blocks for precisely these reasons. But like you said there is this feeling that static blocks should be the default easier choice for most. 

The reason why I created this discussion is mainly to see how it would even be possible to deal with deprecated versions of directives. Is it possible to conditionally generate different `stores` depending on which markup version you currently run into?

**Reply by luisherranz** [reply]
> Is it possible to conditionally generate different stores depending on which markup version you currently run into?

That's interesting 🙂 

We haven't considered it yet, but I guess that as long as you enqueue files manually, you can use the `render_block` filter to inspect the HTML and decide which of your `view.js` versions to enqueue. If this is a recurring need, maybe we can look for ways to automate it somehow.

What do you think?

**Reply by gziolo** [reply]
Aside: once you register `render`/`render_callback` then WordPress sees your block as dynamic and you can check that with the [`WP_Block_Type::is_dynamic()`](https://developer.wordpress.org/reference/classes/wp_block_type/is_dynamic/) method. In theory, you could use that in the WP hooks like [`render_block`](https://developer.wordpress.org/reference/hooks/render_block/) to add some additional functionality.

**Reply by gaambo** [reply]
> There is a sense in the WordPress community that creating static blocks is the default and less advanced way, but that path leads to frustrations when people start discovering their shortcomings. We should revert that trend and start treating dynamic blocks as the default and beginner friendly. Then, teach them that static blocks exist for a specific use case.

I'd like to add to this: I'm not even sure if it's from the documentation or just many of the tutorials on the web. But yes, there's a feeling that a static block should be the default and dynamic block only be used when you have "real" dynamic data (eg posts to fetch, apis to call,...). I think dynamic blocks should be first class citizens and probably the default mode for most developers. 

> The developer experience of adding a directive using the HTML Tag Processor is not as good as adding it directly in the markup.  

That is absolutely the case: I'm happy the HTML Tag Processor made it into core, so it's easier to extend core and 3rd party blocks. But I think using templates and encouraging to put actions/filters in their templates or allow themes to overwrite them (eg like woocommerce templates) would be the preferred way. 

Coming back to the topic of **directives**: I just wanted to +1 everything @fabiankaegy said. 
I'm not yet fully sold on the directives API - but I will add my comments on the blog post/other discussions.
